### PR TITLE
Fix invalid fusion of Matmul and Mul

### LINKED ISSUE #39572: The New Converter of Tensorflow 2.2.0 generate incorrect model output
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): windows/ubuntu
- TensorFlow installed from (source or binary): pip
- TensorFlow version (or github SHA if from source): 2.2.0

**Command used to run the converter or code if you’re using the Python API**

``` python
https://colab.research.google.com/drive/1uMiHGTFrj7R_rFcGJoje3Yfbap0sC9H-?usp=sharing
import tensorflow as tf
import numpy as np
np.set_printoptions(suppress=True)

length = 66

a = tf.constant(
    np.random.rand(length,length).astype(np.float32),
    shape=[length, length])

c = [ tf.constant(
    np.random.rand(length).astype(np.float32),
    shape=[ length]) for i in range(0,3)]


@tf.function
def func2(x ,c1, c2):
    return  tf.multiply(x, c2) + c2

@tf.function
def func(x):
    return  func2(tf.nn.bias_add(tf.matmul(x,a) , c[0] ), c[1],  c[2])
     


input =  np.random.rand(1,length).astype(np.float32)
original_output = func(input).numpy()

print("Orginal:")
print(original_output)


def save_and_test(experimental_new_converter , filename):

    lite = tf.lite.TFLiteConverter.from_concrete_functions([func.get_concrete_function(x = tf.TensorSpec(shape=[None,length], dtype=tf.float32) )])
    lite.experimental_new_converter  = experimental_new_converter
    open(filename,"wb").write(lite.convert())

    interpreter = tf.lite.Interpreter(model_path=filename)
    interpreter.allocate_tensors()
    interpreter.set_tensor(interpreter.get_input_details()[0]['index'], input)
    interpreter.invoke()

    print(filename, ":")
    output = interpreter.get_tensor(interpreter.get_output_details()[0]['index'])
    print(output)
    print(filename, " sum of absolute difference: ", np.sum(np.absolute(original_output - output)))


save_and_test(True, "mlir")
save_and_test(False, "toco")
```

**The output from the converter invocation**

``` 
Orginal:
[[ 6.9496317   7.9791102   0.23810546  1.6724037   9.687994   11.553456
   0.60509163  8.606714   17.731003    2.52809    16.075905    0.50299734
   6.119868   12.388639    8.383967   11.989329    7.3116794   8.7486725
   9.714089   12.450738    1.4212162   7.567941   14.836599    5.73806
   8.028838    2.5774102   7.9609547  19.35534    11.474097    1.2807347
   5.075203   17.249842    3.0866256  17.775652    7.390017   15.896519
  16.136719    6.5862875  12.216141    0.33137095 18.97356     1.9778614
   5.3641543   1.9284656   1.8500257  16.767817   12.7209     14.047361
   7.367909    3.1752422   5.949711   15.742522    2.2939441  16.137108
   1.6687201  13.069997    8.666047   16.782955    1.3117387   6.3798137
   6.557558   15.286772    7.5772853   9.168247    3.2267199   1.5568383 ]]

mlir :
[[ 8.160242   9.0523     7.3261642  8.867854   7.860682   8.722304
   7.2230916  8.622743   9.271859   7.7777925  9.489797   7.390472
   8.524878   9.906944   8.355099   8.330111   8.341421   8.271743
   8.937096   8.515015   7.225615   9.190289   8.9844675  8.162167
   8.694139   6.9362564  9.618863  10.1376915  8.462459   8.29322
   8.158456   8.796178   8.039083   9.774652   8.209907   9.378836
  10.289057   7.984726   8.897137   9.119129   9.802201   8.349303
   6.498592   8.638722   8.569397   9.730883   8.358421   9.018978
   8.4079275  6.496749   8.111706   8.541293   9.513434   9.01988
   7.2715945  8.307488   8.566431   9.748925   8.775068   7.9520493
   7.9917192  9.261906   7.9303417  9.471784   9.427535   7.840794 ]]
mlir  sum of absolute difference:  284.37515

toco :
[[ 6.9496307   7.9791093   0.23810542  1.6724037   9.687995   11.553459
   0.6050917   8.606714   17.731003    2.5280902  16.075907    0.50299734
   6.1198673  12.388639    8.383966   11.98933     7.311679    8.748673
   9.714088   12.450737    1.4212162   7.567941   14.836597    5.73806
   8.028838    2.5774097   7.960953   19.355337   11.474101    1.2807347
   5.075203   17.24984     3.086626   17.775652    7.390019   15.896517
  16.13672     6.586289   12.216139    0.33137095 18.973562    1.9778614
   5.364155    1.9284654   1.8500254  16.767818   12.7209015  14.047361
   7.367908    3.1752434   5.94971    15.74252     2.2939441  16.137106
   1.6687204  13.069999    8.666047   16.782953    1.3117385   6.379814
   6.557558   15.286772    7.577286    9.168246    3.2267199   1.5568382 ]]
toco  sum of absolute difference:  5.4582953e-05
```

**Failure details**
The conversion is successful, but the generated model is wrong,
The result running with the model is much different from the original result.

### PR DESCRIPTION
**Pull request by lgeiger**
This PR fixes #39572 which lead to wrong operator fusion in the TFLite converter for some cases.

This is a resubmission of #40013 which has already be reviewed in detail there.

@rthadur @liufengdb could you take a look?

### CODE REVIEW

**Comment by lgeiger** [comment]
@liufengdb Could you please take a look?

**Comment by lgeiger** [comment]
Any updates on this?

**Comment by lgeiger** [comment]
@joker-eph @liufengdb @karimnosseir @gbaned Any news on when this PR can get a review? I think this PR fixes a pretty serious bug in TFLite so it would be great if this could make it into the 2.4 release.

**Comment by lgeiger** [comment]
Any updates on this? The main changes of this PR have already been reviewed in #40013 and discussed in #39572 which is a major issue for people using fully connected layers with batch normalisation in TFLite.

**Comment by gbaned** [comment]
@liufengdb @karimnosseir Any update on this PR? Please. Thanks!

**Comment by lgeiger** [comment]
Any news?

**Comment by lgeiger** [comment]
@mihaimaruseac Sorry to ping you here, it looks like this PR is stuck in the pipeline.
It fixes #39572 which has been a serious issue for us since TF 2.2 and is a resubmission of #40013 which has been abandoned but already reviewed in detail.
It would be really good to get this reviewed and merged into 2.4 before the release branch get's cut.

**Comment by karimnosseir** [comment]
That got lost for me. I am sorry. I think this pattern was added by @liufengdb, so deferring to him to review.
@lgeiger if got delayed more, ping me and i will check.
Apologies again.

**Comment by lgeiger** [comment]
Thanks for the response, looking forward to the review.

**Comment by lgeiger** [comment]
@karimnosseir For reference, @liufengdb reviewed the content of this PR in #40013 which wasn't merged since https://github.com/tensorflow/tensorflow/pull/40013#discussion_r452659910 wasn't addressed and the PR abandoned by the original author.

**Comment by lgeiger** [comment]
@karimnosseir Any updates?

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc [hunk 1]
```diff
@@ -372,26 +372,26 @@ struct FuseFullyConnectedAndMul : public OpRewritePattern<TFL::MulOp> {
...
-      if (!IsBroadcastableElementsAttrAndType(new_type, filter.getType())) {
-        return failure();
-      }
-      auto new_op =
-          rewriter.create<ConstantOp>(mul_op.getLoc(), new_type, new_cst);
-      new_const_val = new_op.getResult();
+    // Check if it is possible to fuse the constant mathematically
+    // output[..., b, j] = (sum_i input[..., b, i] * filter[j, i]) * m[..., b,
+    // j] m[..., b, j] must be independent of ..., b in order to fuse into
+    // filter
```
Comment (karimnosseir) [inline tensorflow/compiler/mlir/lite/transforms/optimize.cc:378]: It will be good if you can expand the comment with more details. By explaining the requirements, why and how it is done.

File: tensorflow/compiler/mlir/lite/utils/validators.h [hunk 2]
```diff
@@ -70,6 +70,8 @@ inline bool TFPaddingIsSameOrValid(Operation *op, StringAttr *padding) {
 /// Returns whether the given `a` and `b` have broadcast-compatible
 /// types.
 bool IsBroadcastableElementsAttrs(mlir::Attribute a, mlir::Attribute b);
+bool IsDimensionsDegenerateExceptLastOne(mlir::Attribute val);
```
Comment (karimnosseir) [inline tensorflow/compiler/mlir/lite/utils/validators.h:73]: Add Function comments please.

File: tensorflow/lite/python/lite_v2_test.py [hunk 3]
```diff
@@ -1192,5 +1192,81 @@ def model(in_tensor):
...
+      concrete_func = root.f.get_concrete_function(input_data)
+
+      # Convert model.
+      converter = lite.TFLiteConverterV2.from_concrete_functions(
+          [concrete_func])
+      tflite_model = converter.convert()
+
+      # Check values from converted model.
+      expected_value = root.f(input_data)
+      converter.experimental_new_converter = True
```
Comment (karimnosseir) [inline tensorflow/lite/python/lite_v2_test.py:1255]: You can remove this line - it is the default.

File: tensorflow/lite/python/lite_v2_test.py [hunk 4]
```diff
@@ -1192,5 +1192,81 @@ def model(in_tensor):
         str(error.exception))
 
 
+class AffineOpThenMulFusionTest(lite_v2_test_util.ModelTest):
```
Comment (karimnosseir) [inline tensorflow/lite/python/lite_v2_test.py:1195]: func1/func2/... are very general names. Please use more descriptive names to what it is testing, and add comments in needed parts please.

Thanks

File: tensorflow/lite/python/lite_v2_test.py [hunk 5]
```diff
@@ -1192,5 +1192,81 @@ def model(in_tensor):
...
+
+      # Check values from converted model.
+      expected_value = root.f(input_data)
+      converter.experimental_new_converter = True
+
+      actual_value = self._evaluateTFLiteModel(tflite_model, [input_data])
+      np.testing.assert_almost_equal(expected_value.numpy(), actual_value[0])
+
+      # Enable hybrid quantization, same result
+      converter.experimental_new_converter = True
```
Comment (karimnosseir) [inline tensorflow/lite/python/lite_v2_test.py:1261]: duplicate and not needed. Remove please

**Review by karimnosseir — CHANGES_REQUESTED** [review]
Generally LG.
Left few comments.

Thanks

File: tensorflow/lite/python/lite_v2_test.py (hunk 5 above)
Comment (lgeiger) [reply tensorflow/lite/python/lite_v2_test.py:1261]: Good catch, this code is still from the time when it wasn't enabled by default.

File: tensorflow/lite/python/lite_v2_test.py (hunk 4 above)
Comment (lgeiger) [reply tensorflow/lite/python/lite_v2_test.py:1195]: I refactored the tests to make them clearer and expanded the test coverage.

**Comment by lgeiger** [comment]
@karimnosseir Thanks for the review. I rebased the PR onto latest master in b72545ee21dfdc1fde0424a87523536c353f3ce3 and clarified the comments and cleaned up the unittests according the the review suggestions.

Sorry about some of the unclear descriptions, I adapted most of the code from #40013 without too much thinking. I hope this is now clearer to read.

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc [hunk 6]
```diff
@@ -438,26 +438,26 @@ struct FuseFullyConnectedAndMul : public OpRewritePattern<TFL::MulOp> {
...
-          normalized_shape, cst.getType().getElementType()));
-      Type new_type = new_cst.getType();
-      if (!IsBroadcastableElementsAttrAndType(new_type, filter.getType())) {
-        return failure();
-      }
-      auto new_op =
-          rewriter.create<ConstantOp>(mul_op.getLoc(), new_type, new_cst);
-      new_const_val = new_op.getResult();
+    // Only fuse multiplier if all dimensions other than the depth dimension
+    // are equal to 1.
```
Comment (karimnosseir) [inline tensorflow/compiler/mlir/lite/transforms/optimize.cc:442]: Expand comment on why we only fuse in this case.

**Review by karimnosseir — APPROVED** [review]

**Review by karimnosseir — APPROVED** [review]
Thanks

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc (hunk 6 above)
Comment (lgeiger) [reply tensorflow/compiler/mlir/lite/transforms/optimize.cc]: Thanks for the reply, I tried to improve the comment in 2b2b3a9bd817b05e2451b2de03f170b3282d63f5.

The reason for this is that we can only fuse a multiplier in dimensions that are not summed over during the matmul due to:
Screenshot 2020-10-20 at 18 07 56

**Review by gbaned — APPROVED** [review]

**Comment by lgeiger** [comment]
@gbaned looks like there are quite a few internal CI failures. Unfortunately I cannot access the logs, is there anything that is still preventing this from being merged from my side?

**Comment by gbaned** [comment]
@lgeiger  It is processing internally, we will let you know if anything required from you. Thank you!

**Comment by lgeiger** [comment]
@gbaned any updates state of the internal processing?

File: tensorflow/compiler/mlir/lite/utils/validators.cc [hunk 7]
```diff
@@ -78,5 +78,19 @@ bool IsBroadcastableElementsAttrs(mlir::Attribute a, mlir::Attribute b) {
...
+bool IsDimensionsDegenerateExceptLastOne(ArrayRef<int64_t> elements_shape) {
+  for (auto i = 0; i < elements_shape.size() - 1; i++) {
+    if (elements_shape[i] != 1) return false;
+  }
+  return true;
+}
+
+bool IsDimensionsDegenerateExceptLastOne(Attribute val) {
+  if (auto elements = val.dyn_cast<DenseElementsAttr>()) {
+    return IsDimensionsDegenerateExceptLastOne(elements.getType().getShape());
```
Comment (abattery) [inline tensorflow/compiler/mlir/lite/utils/validators.cc:90]: This line should be rewritten in order to access the shape information if available.

For example,

if (auto ranked_type = elements.getType().cast<RankedTensorType>()) {
  return IsDimensionsDegenerateExceptLastOne(ranked_type.getShape());
}

Otherwise, it will cause crashing when the type is UnrankedTensorType.

Comment (lgeiger) [reply tensorflow/compiler/mlir/lite/utils/validators.cc]: Good catch, I fixed it in 4ebd2633e870dd474fbc9b348d311a4ccdf20158

**Review by abattery — APPROVED** [review]

File: tensorflow/compiler/mlir/lite/utils/validators.cc [hunk 8]
```diff
@@ -78,5 +78,19 @@ bool IsBroadcastableElementsAttrs(mlir::Attribute a, mlir::Attribute b) {
...
 
+bool IsDimensionsDegenerateExceptLastOne(ArrayRef<int64_t> elements_shape) {
+  for (auto i = 0; i < elements_shape.size() - 1; i++) {
+    if (elements_shape[i] != 1) return false;
+  }
+  return true;
+}
+
+bool IsDimensionsDegenerateExceptLastOne(Attribute val) {
+  if (auto ranked_type = val.getType().cast<RankedTensorType>()) {
```
Comment (abattery) [inline tensorflow/compiler/mlir/lite/utils/validators.cc:89]: Sorry for my bad. Let's use dyn_cast to make sure avoid crashing on UnrankedTensorType.

```
if (auto ranked_type = val.getType().dyn_cast<RankedTensorType>()) {
    return IsDimensionsDegenerateExceptLastOne(ranked_type.getShape());
}
```

File: tensorflow/compiler/mlir/lite/utils/validators.cc [hunk 9]
```diff
@@ -78,5 +78,19 @@ bool IsBroadcastableElementsAttrs(mlir::Attribute a, mlir::Attribute b) {
   return OpTrait::util::getBroadcastedType(a.getType(), b.getType()) != Type();
 }
 
+bool IsDimensionsDegenerateExceptLastOne(ArrayRef<int64_t> elements_shape) {
+  for (auto i = 0; i < elements_shape.size() - 1; i++) {
```
Comment (abattery) [inline tensorflow/compiler/mlir/lite/utils/validators.cc:82]: zero valued size_t type minus - 1 is problematic and it is causing test failures.

Let's use the ArrayRef member methods as much as possible to avoid unexpected crashing due to the type conversion:

```
  if (elements_shape.empty()) return true;

  for (auto dim : elements_shape.drop_back(1)) {
    if (dim != ShapedType::kDynamicSize) return false;
  }
  return true;
```

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc [hunk 10]
```diff
@@ -438,26 +438,28 @@ struct FuseFullyConnectedAndMul : public OpRewritePattern<TFL::MulOp> {
...
-          rewriter.create<ConstantOp>(mul_op.getLoc(), new_type, new_cst);
-      new_const_val = new_op.getResult();
+    // Only fuse multiplier if all dimensions other than the depth dimension
+    // are equal to 1 since otherwise
+    // `matmul(x, filter) * cst != matmul(x, filter * cst)`
+    // even if `filter` and `cst` are be broadcastable.
+    auto shape = cst.getType().getShape();
+    if (!IsDimensionsDegenerateExceptLastOne(shape)) return failure();
+
+    int64_t element_size = shape.size() > 0 ? shape[shape.size() - 1] : 1;
```
Comment (abattery) [inline tensorflow/compiler/mlir/lite/transforms/optimize.cc:448]: Comparing size() with zero can be easily replaced with empty().

```
    int64_t element_size = shape.empty() ? 1 : shape[shape.size() - 1];
```

File: tensorflow/compiler/mlir/lite/utils/validators.cc (hunk 9 above)
Comment (abattery) [reply tensorflow/compiler/mlir/lite/utils/validators.cc]: I confused about 1 and -1. Here is the updated suggestion.

```
  if (elements_shape.empty()) return true;

  for (auto dim : elements_shape.drop_back(1)) {
    if (dim != 1) return false;
  }
  return true;
```

Comment (lgeiger) [reply tensorflow/compiler/mlir/lite/utils/validators.cc]: Thanks for the review. I addressed the comments in ce4e534ad61012b724b80d1aab2f68017d214e76 however I didn't use `kDynamicSize` is it would change behaviour since we need the dimensions to be `1` and not dynamic, or am I missing something?

**Review by abattery — APPROVED** [review]

**Comment by abattery** [comment]
@lgeiger could you resolve the merge conflict?

**Comment by abattery** [comment]
@lgeiger Nevermind. I will handle internal reviews.
